### PR TITLE
`ItemListener.onCheckDelete` & `ItemDeletion.cancelBuildsInProgress`

### PR DESCRIPTION
**Pull request by jglick**
In CloudBees CI running in HA mode we need to augment the current behavior of `AbstractItem.delete` & `ItemDeletion`, which aborts running builds of a job before it is deleted, to accommodate multiple replicas. `ItemListener.onDeleted` is not helpful because it is called _after_ the job has been deleted, by which point it is too late to terminate any running builds gracefully. Additionally, the somewhat complex implementation of the abort section of the `delete` method (the part prior to that patched in #8645) makes the whole method very long and hard to follow, and would best not be duplicated.

This change introduces an extension to `ItemListener` allowing listeners to be notified of a proposed deletion and either veto it or clean up ongoing processes first, by analogy with an existing callback for copying. It also moves the cancel behavior into a standalone utility method in `ItemDeletion`, since it is closely related to the other queue code there.

(I also looked into making `ItemDeletion` register its own `ItemListener` so it would be self-contained, not referenced from `AbstractItem.delete`. This did not seem practical because the `register` and `deregister` calls need to be paired around the actual deletion, which would have required yet another listener method like `onDeleted` but always called even if the deletion failed. That just seemed like it would be confusing.)

### Testing done

An integration test in CloudBees CI exercises an implementation of the new listener callback, which calls the newly available utility method.

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/AbstractItem.java [hunk 1]
```diff
@@ -705,11 +697,13 @@ public void delete(StaplerRequest req, StaplerResponse rsp) throws IOException,
      *
      * <p>
      * Any exception indicates the deletion has failed, but {@link AbortException} would prevent the caller
-     * from showing the stack trace. This
+     * from showing the stack trace.
```
Comment (jglick) [inline core/src/main/java/hudson/model/AbstractItem.java:700]: https://github.com/jenkinsci/jenkins/commit/ea9ddafac2ddefa3feb5a091dacf045c4f2fa9a5#r8089452

File: core/src/main/java/hudson/model/listeners/ItemListener.java [hunk 2]
```diff
@@ -205,6 +218,19 @@ public static void fireOnUpdated(final Item item) {
         Listeners.notify(ItemListener.class, false, l -> l.onUpdated(item));
     }
 
+    @Restricted(NoExternalUse.class)
+    public static void checkBeforeDelete(Item item) throws Failure, InterruptedException {
+        for (ItemListener l : all()) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/listeners/ItemListener.java:223]: Like `checkBeforeCopy`, does _not_ use `Listeners.notify`, and runs with the same authn as `delete`.

**Review by Vlatombe — APPROVED** [review]

File: core/src/main/java/hudson/model/listeners/ItemListener.java [hunk 3]
```diff
@@ -94,6 +96,17 @@ public void onCopied(Item src, Item item) {
     public void onLoaded() {
     }
 
+    /**
+     * Called before an item is deleted, providing the ability to veto the deletion operation before it starts.
+     * @param item the item being deleted
+     * @param parent the proposed parent
```
Comment (Vlatombe) [inline core/src/main/java/hudson/model/listeners/ItemListener.java:102]:
```suggestion
```

**Review by julieheard — APPROVED** [review]

**Review by MarkEWaite — COMMENTED** [review]
This PR is now ready for merge.  We will merge it after approximately 24 hours if there is no negative feedback.

/label ready-for-merge
